### PR TITLE
APIGOV-22589

### PR DESCRIPTION
**Pull request by tsjohns9**
- Update discovery cache to save resources based on the provided handlers.
- Fetch marketplace provisioning resources and pass them to the handlers in sequential order.
- Register handlers based on the agent type

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -423,19 +432,28 @@ func newHandlers() []handler.Handler {
...
 		agent.proxyResourceHandler,
 	}
 
+	if agent.cfg.GetAgentType() == config.DiscoveryAgent {
+		handlers = append(
+			handlers,
+			handler.NewCategoryHandler(agent.cacheManager),
+			handler.NewCRDHandler(agent.cacheManager),
+			handler.NewARDHandler(agent.cacheManager),
+			// handler.NewACLHandler(agent.cacheManager),
```
Comment (tsjohns9) [inline pkg/agent/agent.go:445]: @jcollins-axway is this still needed?

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -423,19 +432,28 @@ func newHandlers() []handler.Handler {
...
 		handler.NewInstanceHandler(agent.cacheManager),
-		handler.NewCategoryHandler(agent.cacheManager),
 		handler.NewAgentResourceHandler(agent.agentResourceManager),
-		handler.NewCRDHandler(agent.cacheManager),
-		handler.NewARDHandler(agent.cacheManager),
-		// handler.NewACLHandler(agent.cacheManager),
 		agent.proxyResourceHandler,
 	}
 
+	if agent.cfg.GetAgentType() == config.DiscoveryAgent {
```
Comment (tsjohns9) [inline pkg/agent/agent.go:439]: Only registering handlers based on what is defined in the watch topic for the agent type.

File: pkg/agent/agent.go (hunk 1 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:445]: I kept it to enable again when we do fix the update handling of ACLs on API Server

Comment (sbolosan) [inline pkg/agent/agent.go:445]: @jcollins-axway  this is still here for future use, yeah?

**Review by sbolosan — DISMISSED** [review]

File: pkg/agent/agent.go (hunk 1 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:445]: yup

**Review by sbolosan — APPROVED** [review]

File: pkg/agent/discoverycache.go [hunk 3]
```diff
@@ -24,264 +25,428 @@ const (
...
+		return nil
+	}
+
+	dc.logger.WithField("kind", "Credential").Trace("fetching Credentials and updating cache")
+
+	credLink := mv1.NewCredential("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(credLink)
+
+	credentials, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		nil, url, apiServerPageSize,
```
Comment (tsjohns9) [inline pkg/agent/discoverycache.go:404]: @jcollins-axway Are there certain fields to query for on the credential like we do for other resources?

File: pkg/agent/discoverycache.go [hunk 4]
```diff
@@ -24,264 +25,428 @@ const (
...
 
-func (j *discoveryCache) updateCategoryCache() {
-	j.logger.Trace("updating category cache")
+func (dc *discoveryCache) handleAccessControlList() error {
+	dc.logger.WithField("kind", "AccessControlList").Trace("fetching AccessControlList and updating cache")
 
 	// Update cache with published resources
-	existingCategories := make(map[string]bool)
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields,
```
Comment (tsjohns9) [inline pkg/agent/discoverycache.go:258]: @jcollins-axway Are these the right fields to query for on an access control list?

File: pkg/agent/discoverycache.go [hunk 5]
```diff
@@ -24,264 +25,428 @@ const (
...
 	)
+	if err != nil {
+		e := utilErrors.Wrap(ErrUnableToGetAPIV1Resources, err.Error()).FormatError("APIServiceInstances")
+		dc.logger.Error(e)
+		return e
+	}
+
+	for _, instance := range serviceInstances {
+		if err := dc.handleResource(instance); err != nil {
+			return err
```
Comment (tsjohns9) [inline pkg/agent/discoverycache.go:221]: @jcollins-axway for all of these funcs I am returning an error if the handler failed to process the resource. Maybe that isn't necessary and we should just log the error if that happens. I don't think everything needs to stop. What do you think?

**Review by vivekschauhan — APPROVED** [review]

File: pkg/agent/discoverycache.go (hunk 5 above)
Comment (jcollins-axway) [reply pkg/agent/discoverycache.go:221]: I agree, I think we need to look at these situations in the migration code as well. Was thinking about this after talking to Shane about the issue Dave hit.

File: pkg/agent/discoverycache.go (hunk 3 above)
Comment (jcollins-axway) [reply pkg/agent/discoverycache.go:404]: I do not think we should limit the fields we get.

**Review by jcollins-axway — DISMISSED** [review]

File: pkg/agent/discoverycache.go (hunk 4 above)
Comment (tsjohns9) [reply pkg/agent/discoverycache.go:258]: @jcollins-axway Can we remove the query for all the resources we fetch then?

**Review by sbolosan — DISMISSED** [review]

**Review by jcollins-axway — APPROVED** [review]

**Review by sbolosan — APPROVED** [review]

File: pkg/agent/events/watchtopic.go [hunk 6]
```diff
@@ -65,8 +65,8 @@ var (
 
 // getOrCreateWatchTopic attempts to retrieve a watch topic from central, or create one if it does not exist.
 func getOrCreateWatchTopic(name, scope string, client APIClient, features watchTopicFeatures) (*mv1.WatchTopic, error) {
-	wt := mv1.NewWatchTopic(name)
-	ri, err := client.GetResource(wt.GetSelfLink())
+	wt := mv1.NewWatchTopic("")
+	ri, err := client.GetResource(fmt.Sprintf("%s/%s", wt.GetKindLink(), name))
```
Comment (jcollins-axway) [inline pkg/agent/events/watchtopic.go:69]: This seems to be a 1 for 1 change, is there a benefit of one way vs the other?

**Review by sbolosan — DISMISSED** [review]

**Review by sbolosan — APPROVED** [review]

**Review by jcollins-axway — APPROVED** [review]

**Pull request merged**